lib/devise_security_extension/hooks/session_lockable.rb

```diff
@@ -1,33 +1,22 @@
-Warden::Manager.after_authentication do |record, warden, options|
-  if record.respond_to?(:login_attempts)
-    record.update(login_attempts: 0, most_recent_attempt_at: nil)
-  end
-end
+Warden::Manager.before_failure do |env, warden, options|
+  request_parameters = env['action_dispatch.request.request_parameters']
+  user_parameters = request_parameters['user'] if request_parameters
+  email = user_parameters['email'] if user_parameters
+  scope = warden[:scope].to_s.classify.constantize if warden[:scope]
+  record = scope.where(email: email).first if email && scope
 
-Warden::Manager.before_failure do |record, warden, options|
-  if record.respond_to?(:login_attempts)
-    if record.advanced_security_required?
-      record.update(login_attempts: record.login_attempts + 1, most_recent_attempt_at: Time.now)
-    end
-  else
-    request_parameters = record['action_dispatch.request.request_parameters']
-    user_parameters = request_parameters['user'] if request_parameters
-    email = user_parameters['email'] if user_parameters
-    user = User.where(email: email).first if email
-
-    if user
-      if user.advanced_security_required?
-        user.login_attempts += 1
-        user.most_recent_attempt_at = Time.now
-        user.save
-      end
-    end
+  if record && record.advanced_security_required?
+    record.update_attribute(:login_attempts, record.login_attempts + 1)
+    record.update_attribute(:most_recent_attempt_at, Time.now)
   end
 end
 
-Warden::Manager.after_set_user do |user, auth, opts|
-  if user.login_attempts > 6 and user.most_recent_attempt_at >= 3.minutes.ago
+Warden::Manager.after_set_user do |record, auth, opts|
+  if record.login_attempts > 6 && record.most_recent_attempt_at >= 3.minutes.ago
     auth.logout
-    throw :warden, message: "Your account has been locked out. Please confirm your password and try back later."
+    throw :warden, message: 'Your account has been locked out. Please confirm your password and try back later.'
```

Comment:
nice! Much easier to follow. Do we want to notify the user immediately when they're locked out so? I think that's how other login sites behave and it would avoid frustration from users spending a lot of time trying to login only to find out they're locked out once they finally figured out the password.